collect_app/src/main/java/org/odk/collect/android/application/initialization/ApplicationInitializer.kt

```diff
@@ -120,3 +191,19 @@ class ApplicationInitializer(
         }
     }
 }
+
+class ContainsSearchOrPullDataExtra(containsPullData: Boolean) : Externalizable {
+
+    constructor() : this(false)
+
+    var containsSearchOrPullData = containsPullData
+        private set
+
+    override fun readExternal(`in`: DataInputStream?, pf: PrototypeFactory?) {
```

Comment:
On upgrade, this will fail, form loading from the cached representation will throw an exception, and form loading from XML will be invoked?

Reply:
I had missed this while dealing with your other comments. There is actually a problem here, but it's not to do with `readExternal`. Cached versions of the form never had `DynamicPreloadExtra` added to their `FormDef`, so it will never be read from the external representation. However, there is code in Collect now that assumes that a `FormDef` will have a `DynamicPreloadExtra` and we'll most likely see a crash/parse error there. I'll look into it.

Reply:
Ok fixed that in https://github.com/getodk/collect/pull/5665/commits/2ca6909068a9ac6dd24279add7f394861e34c147. 

I was then very confused as to why we didn't run into similar problems with entities as it looks like we also assume `FormDef` instances have a `EntityFormExtra` in `EntityFormFinalizationProcessor`. My guess is that because we added the concept of "extras" for `FormDef` in that release, cached `FormDef` instances would fail to deserialize as they wouldn't have the `ExternalizableExtras` included as part of them and then be deleted (silently). There's definitely a scary realisation in there for me that we don't have any versioning built (like we get from something like `Serializable`), so it's pretty easy to miss potential upgrade problems like this (and manually testing they don't exist is subtle). I'll add this as a conversation topic.